save.go

```diff
@@ -138,50 +140,53 @@ func (p byImportPath) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
-	var filteredCmds []string
-	for _, cmd := range cmds {
-		if !strings.HasPrefix(cmd, importPath) {
-			filteredCmds = append(filteredCmds, cmd)
+	roots := make(map[string]struct{})
+	deps := make(map[string]struct{})
+	for _, useAllFiles := range []bool{false, true} {
+		buildContext := build.Default
+		buildContext.CgoEnabled = true
+		buildContext.UseAllFiles = useAllFiles
+
+		forward, _, errs := importgraph.Build(&buildContext)
+		for pkg, err := range errs {
+			// Lots of errors because of UseAllFiles.
+			if !useAllFiles {
+				log.Printf("error loading package %s: %v", pkg, err)
+			}
 		}
-	}
 
-	// Get a set of transitive dependencies (package import paths) for the
-	// specified package.
-	var pkgExprs = append([]string{path.Join(importPath, "...")}, filteredCmds...)
-	var output = mustRun("go",
-		append([]string{"list", "-f", `{{range .Deps}}{{.}}{{"\n"}}{{end}}`}, pkgExprs...)...)
-	var deps = filterPackages(output, nil) // filter out standard library
-
-	// Add the command packages.
-	for _, cmd := range filteredCmds {
-		deps[cmd] = struct{}{}
-	}
+		// Add the command packages as both dependencies and roots.
+		for _, pkg := range cmds {
+			deps[pkg] = struct{}{}
+			roots[pkg] = struct{}{}
+		}
 
-	// List dependencies of test files, which are not included in the go list .Deps
-	// Also, ignore any dependencies that are already covered.
-	var testImportOutput = mustRun("go",
-		append([]string{"list", "-f", `{{join .TestImports "\n"}}{{"\n"}}{{join .XTestImports "\n"}}`}, pkgExprs...)...)
-	var testImmediateDeps = filterPackages(testImportOutput, deps) // filter out stdlib and existing deps
-	for dep := range testImmediateDeps {
-		deps[dep] = struct{}{}
-	}
+		// Get all subpackages.
+		for pkg := range forward {
+			if strings.HasPrefix(pkg, importPath) {
+				roots[pkg] = struct{}{}
+			}
+		}
 
-	// We have to get the transitive deps of the remaining test imports.
-	// NOTE: this will return the dependencies of the libraries imported by tests
-	// and not imported by main code.  This output does not include the imports themselves.
-	if len(testImmediateDeps) > 0 {
-		var testDepOutput = mustRun("go", append([]string{"list", "-f", `{{range .Deps}}{{.}}{{"\n"}}{{end}}`}, setToSlice(testImmediateDeps)...)...)
-		var allTestDeps = filterPackages(testDepOutput, deps) // filter out standard library and existing deps
-		for dep := range allTestDeps {
-			deps[dep] = struct{}{}
+		// Get the reflexive transitive closure for all packages of interest.
+		for pkg := range forward.Search(setToSlice(roots)...) {
+			deps[pkg] = struct{}{}
 		}
 	}
 
-	// Return everything in deps
 	var result []string
-	for dep := range deps {
-		result = append(result, dep)
+	for pkg := range deps {
+		// Exclude our roots. Note that commands are special-cased above.
+		if _, ok := roots[pkg]; ok {
+			continue
+		}
+		slash := strings.IndexByte(pkg, '/')
+		stdLib := slash == -1 || strings.IndexByte(pkg[:slash], '.') == -1
```

Comment:
For my understanding, anything not in the standard library will be a relative path?